README.md
# PI Web API Python Sample

**Version:** 1.0.33

[![Build Status](https://dev.azure.com/osieng/engineering/_apis/build/status/product-readiness/PI-System/aveva.sample-pi_web_api-common_actions-python?branchName=main)](https://dev.azure.com/osieng/engineering/_build/latest?definitionId=2663&branchName=main)

The sample code in this folder demonstrates how to utilize the PI Web API using Python. You must have already [installed Python](https://www.python.org/downloads/) in order to run this sample application.

## Getting Started

To run the sample code:

- Clone the GitHub repository
- Open the "sample-pi_web_api-common_actions-python" folder with your IDE
- Install the required modules by running the following command in the terminal: `pip install -r requirements.txt`
- Run the application using the following command in the terminal: `python .\program.py` where `program.py` is the program you want to run. e.g `python .\create_sandbox.py`

## Getting Started with Tests

To run the sample tests:

- The sample test is configured using the file [appsettings.placeholder.json](appsettings.placeholder.json). Before editing, rename this file to `appsettings.json`. This repository's `.gitignore` rules should prevent the file from ever being checked in to any fork or branch, to ensure credentials are not compromised.
- Open the test config file `appsettings.json`
- Replace the values with your system configuration.

For example:

```json
{
  "Resource": "PLACEHOLDER_REPLACE_WITH_PI_WEB_API_URL",
  "DataArchiveName": "PLACEHOLDER_REPLACE_WITH_DATA_ARCHIVE_NAME",
  "AFServerName": "PLACEHOLDER_REPLACE_WITH_AF_SERVER_NAME",
  "Username": "PLACEHOLDER_REPLACE_WITH_USERNAME",
  "Password": "PLACEHOLDER_REPLACE_WITH_PASSWORD",
  "AuthType": "kerberos",
  "VerifySSL": true
}
```

| Parameters      | Required | Type    | Description                                                                                               |
| --------------- | -------- | ------- | --------------------------------------------------------------------------------------------------------- |
| Resource        | required | string  | The URL of the PI Web API                                                                                 |
| DataArchiveName | required | string  | The name of the PI Data Archive that is being sent to                                                     |
| AFServerName    | required | string  | The name of the AF server that is being sent to                                                           |
| Username        | required | string  | The username that is being used for authenticating to the PI Web API                                      |
| Password        | required | string  | The password that is being used for authenticating to the PI Web API                                      |
| AuthType        | optional | string  | The type of authentication to use when connecting to the PI Web API. By default this is set to "kerberos" |
| VerifySSL       | optional | boolean | A feature flag for verifying SSL when connecting to the PI Web API. By default this is set to true        |

- Each test file (prefixed as "test\_..."), can be run independently or all the tests can be run in a single instance via the `run_all_tests.py` file. **Note: while the tests can be ran individually, some test database, elements, templates, and attributes created within other tests or by the create_sandbox.py script. If these structures are not in place ahead of time, the tests will not function as intended and the API requests will likely return a 404 error**
- To run a single file, open the test file you wish to run: e.g. `.\test_batch.py`

- In the terminal, navigate to the test files and use the following command to run all the tests: `python .\run_all_tests.py` or to run a test individually: `python .\test_batch_call.py`

## System Configuration

In order to run this sample, you must configure PI Web API with the proper security to:

- Create an AF database
- Create AF categories
- Create AF templates
- Create AF elements with attributes
- Create PI Points associated with element attributes
- Write and read element attributes
- Delete all the above AF/PI Data Archive objects

In addition, PI Web API must be configured to allow CORS as follows:

| Attribute               | Value                                               | Type    |
| ----------------------- | --------------------------------------------------- | ------- |
| CorsExposedHeaders      | Allow,Content-Encoding,Content-Length,Date,Location | String  |
| CorsHeaders             | \*                                                  | String  |
| CorsMethods             | \*                                                  | String  |
| CorsOrigins             | [http://localhost:9876](http://localhost:9876)      | String  |
| CorsSupportsCredentials | True                                                | Boolean |
| DisableWrites           | False                                               | Boolean |

## Functionality

This sample shows basic functionality of the PI Web API, not every feature. The sample is meant to show a basic sample application that uses the PI Web API to read and write data to a PI Data Archive and AF. Tests are also included to verify that the code is functioning as expected.

The functionality included with this sample includes(recommended order of execution):

- Create an AF database
- Create a category
- Create an element template
- Create an element and associate the element's attributes with PI tags where appropriate
- Write a single value to the attribute
- Write 100 values to an attribute
- Perform a Batch (6 steps in 1 call) operation which includes:
  - Get the sample tag
  - Read the sample tag's snapshot value
  - Read the sample tag's last 10 recorded values
  - Write a value to the sample tag
  - Write 3 values to the sample tag
  - Read the last 10 recorded values from the sample tag only returning the value and timestamp
- Return all the values over the last 2 days
- Return timestamp and values over the last 2 days
- Delete the element
- Delete the element template
- Delete the sample database

---

For the main PI Web API Samples landing page [ReadMe](https://github.com/osisoft/OSI-Samples-PI-System/tree/main/docs/PI-Web-API-Docs)  
For the main PI System Samples landing page [ReadMe](https://github.com/osisoft/OSI-Samples-PI-System)  
For the main AVEVA Samples landing page [ReadMe](https://github.com/osisoft/OSI-Samples)


HISTORY.md
# Version History

## 1.0.34 / 2023-10-13

- Migrate Pipelines

## 1.0.33 / 2022-08-17

- Automate dependabot approval and automerge

## 1.0.32 / 2022-08-09

- Add dependabot.yml to automate dependency checks
- Remove weekly tests
- Parameterize pool and demands

## 1.0.31 / 2022-06-20

- Updated dependencies

## 1.0.30 / 2021-11-05

- Changed agent used in pipeline

## 1.0.29 / 2022-01-17

- Updated dependencies

## 1.0.28 / 2021-12-07

- Update pipelines to reference internal analysis templates

## 1.0.27 / 2021-10-13

- Update pipelines to use Azure Key Vault for secrets
- Changed configuration file name and some parameter names

## 1.0.26 / 2021-09-07

- Edited gitignore

## 1.0.25 / 2021-08-10

- Edited gitignore

## 1.0.24 / 2021-07-15

- Updated dependencies

## 1.0.23 / 2021-06-29

- Add option of certificate validation to batch calls

## 1.0.22 / 2021-04-23

- Implement certificate validation
- Configuration file changes

## 1.0.21 / 2021-03-18

- Fixed potential vulnerabilities detected by static analysis

## 1.0.20 / 2021-03-03

- Updated pipeline to reference main branch

## 1.0.19 / 2021-01-29

- Resolved typos in readme

## 1.0.18 / 2021-01-26

- Updated dependencies

## 1.0.17 / 2020-12-16

- Updated dependencies

## 1.0.16 / 2020-09-25

- Fix azure pipelines

## 1.0.15 / 2020-08-06

- Resolve Polaris detected issues

## 1.0.14 / 2020-07-31

- Add sample credentials file to gitignore and replace with placeholder

## 1.0.13 / 2020-07-31

- Updated dependencies

## 1.0.12 / 2020-07-01

- Updated to use Polaris in place of Coverity

## 1.0.11 / 2020-06-10

- Fix badge links in readme

## 1.0.10 / 2020-05-22

- Minor readme cleanup

## 1.0.9 / 2020-03-30

- Updated readme with CorsOrigins info

## 1.0.8 / 2020-02-18

- Add license file for dependencies

## 1.0.7 / 2020-01-28

- Add file listing dependencies

## 1.0.6 / 2020-01-08

- Add always flag to scheduled pipeline build

## 1.0.5 / 2019-12-20

- Add scheduled build for pipeline

## 1.0.4 / 2019-12-17

- Add link to master documentation

## 1.0.3 / 2019-12-13

- Add analysis template to pipeline

## 1.0.2 / 2019-12-06

- Migrate pipelines into new ADO project

## 1.0.1 / 2019-11-05

- Updated pipelines

## 1.0.0 / 2019-08-16

- Initial public release


test_config.py
import json

# Try to open the configuration file
try:
    with open(
        'test_config.json',
        'r',
    ) as f:
        config = json.load(f)
except Exception as error:
    print(f'Error: {str(error)}')
    print(f'Could not open/read file: test_config.json')
    exit()

PIWEBAPI_URL = config.get('Resource')
AF_SERVER_NAME = config.get('AssetServerName')
PI_SERVER_NAME = config.get('DataServerName')
USER_NAME = config.get('Username')
USER_PASSWORD = config.get('Password')
AUTH_TYPE = config.get('AuthType', 'kerberos')
VERIFY_SSL = config.get('VerifySSL', True)



test_batch_call.py
""" This script contains the unit tests for the Batch Calls script """

import unittest

from appsettings import PIWEBAPI_URL, AF_SERVER_NAME, USER_NAME, USER_PASSWORD, AUTH_TYPE, VERIFY_SSL


class TestStringMethods(unittest.TestCase):
    """
        We recommend running these unit tests using the
           file syntax:  python -m unittest test_batch_call
        To discover and run all unit tests:  python -m unittest
        To run the unit tests in a file:  python -m unittest test_batch_call
        To run a single test:  python -m unittest test_batch_call.TestStringMethods.test_dobatchcall
    """

    def test_dobatchcall(self):
        """ Call the do_batch_call method """
        from batch_call import do_batch_call
        self.assertEqual(do_batch_call(PIWEBAPI_URL, AF_SERVER_NAME,
                                       USER_NAME, USER_PASSWORD, AUTH_TYPE, VERIFY_SSL), 207)

if __name__ == '__main__':
    unittest.main()


test_create_sandbox.py
""" This script contains the unit tests for the CreateSandbox Script """

import unittest

from appsettings import PIWEBAPI_URL, AF_SERVER_NAME, PI_SERVER_NAME, USER_NAME, USER_PASSWORD, AUTH_TYPE, VERIFY_SSL


class TestStringMethods(unittest.TestCase):
    """
        Run each unit tests in order, manually using the following example:
            python -m unittest test_create_sandbox.TestStringMethods.test_createdatabase
        To discover and run all unit tests:  python -m unittest
        To run the unit tests in a file:  python -m unittest test_create_sandbox
        To run a single test:
            python -m unittest test_create_sandbox.TestStringMethods.test_createdatabase
    """

    def test_createdatabase(self):
        """ Test the create_database method """
        from create_sandbox import create_database
        self.assertEqual(create_database(PIWEBAPI_URL,
                                         AF_SERVER_NAME, USER_NAME, USER_PASSWORD, AUTH_TYPE, VERIFY_SSL), 201)

    def test_createcategory(self):
        """ Test the create_category method """
        from create_sandbox import create_category
        self.assertEqual(create_category(PIWEBAPI_URL,
                                         AF_SERVER_NAME, USER_NAME, USER_PASSWORD, AUTH_TYPE, VERIFY_SSL), 201)

    def test_createtemplate(self):
        """ Test the create_template method """
        from create_sandbox import create_template
        self.assertEqual(create_template(PIWEBAPI_URL,
                                         AF_SERVER_NAME, PI_SERVER_NAME, USER_NAME, USER_PASSWORD, AUTH_TYPE, VERIFY_SSL), 201)

    def test_createelement(self):
        """ Test the create_element method """
        from create_sandbox import create_element
        self.assertEqual(create_element(PIWEBAPI_URL,
                                        AF_SERVER_NAME, USER_NAME, USER_PASSWORD, AUTH_TYPE, VERIFY_SSL), 200)

    def test_deleteelement(self):
        """ Test the delete_element method """
        from create_sandbox import delete_element
        self.assertEqual(delete_element(PIWEBAPI_URL,
                                        AF_SERVER_NAME, USER_NAME, USER_PASSWORD, AUTH_TYPE, VERIFY_SSL), 204)

    def test_deletetemplate(self):
        """ Test the delete_template method """
        from create_sandbox import delete_template
        self.assertEqual(delete_template(PIWEBAPI_URL,
                                         AF_SERVER_NAME, USER_NAME, USER_PASSWORD, AUTH_TYPE, VERIFY_SSL), 204)

    def test_deletecategory(self):
        """ Test the delete_category method """
        from create_sandbox import delete_category
        self.assertEqual(delete_category(PIWEBAPI_URL,
                                         AF_SERVER_NAME, USER_NAME, USER_PASSWORD, AUTH_TYPE, VERIFY_SSL), 204)

    def test_deletedatabase(self):
        """ Test the create_database method """
        from create_sandbox import delete_database
        self.assertEqual(delete_database(PIWEBAPI_URL,
                                         AF_SERVER_NAME, USER_NAME, USER_PASSWORD, AUTH_TYPE, VERIFY_SSL), 204)


if __name__ == '__main__':
    unittest.main()


read_attributes.py
""" This script read values through the PI Web API.

    This python script requires some pre-requisites:
        1.  A back-end server with PI WEB API with CORS enabled.
"""

import json
import getpass
import requests
from urllib.parse import urlparse

OSI_AF_ATTRIBUTE_TAG = 'OSIPythonAttributeSampleTag'
OSI_AF_DATABASE = 'OSIPythonDatabase'
OSI_AF_ELEMENT = 'OSIPythonElement'


def call_headers(include_content_type):
    """ Create API call headers
        @includeContentType boolean: Flag determines whether or not the
        content-type header is included
    """
    if include_content_type is True:
        header = {
            'content-type': 'application/json',
            'X-Requested-With': 'XmlHttpRequest'
        }
    else:
        header = {
            'X-Requested-With': 'XmlHttpRequest'
        }

    return header


def call_security_method(security_method, user_name, user_password):
    """ Create API call security method
        @param security_method string: Security method to use: basic or kerberos
        @param user_name string: The user's credentials name
        @param user_password string: The user's credentials password
    """
    from requests.auth import HTTPBasicAuth
    from requests_kerberos import HTTPKerberosAuth

    if security_method.lower() == 'basic':
        security_auth = HTTPBasicAuth(user_name, user_password)
    else:
        security_auth = HTTPKerberosAuth(mutual_authentication='REQUIRED',
                                         sanitize_mutual_error_response=False)

    return security_auth


def read_attribute_snapshot(piwebapi_url, asset_server, user_name, user_password,
                            piwebapi_security_method, verify_ssl):
    """ Read a single value
        @param piwebapi_url string: The URL of the PI Web API
        @param asset_server string: Name of the Asset Server
        @param user_name string: The user's credentials name
        @param user_password string: The user's credentials password
        @param piwebapi_security_method string: Security method: basic or kerberos
        @param verify_ssl: If certificate verification will be performed
    """
    print('readAttributeSnapshot')

    #  create security method - basic or kerberos
    security_method = call_security_method(
        piwebapi_security_method, user_name, user_password)

    #  Get the sample tag
    request_url = '{}/attributes?path=\\\\{}\\{}\\{}|{}'.format(
        piwebapi_url, asset_server, OSI_AF_DATABASE, OSI_AF_ELEMENT, OSI_AF_ATTRIBUTE_TAG)
    response = requests.get(request_url, auth=security_method, verify=verify_ssl)

    #  Only continue if the first request was successful
    if response.status_code == 200:
        print(response.text)
        #  Deserialize the JSON Response
        data = json.loads(response.text)

        url = urlparse(piwebapi_url + '/streams/' + data['WebId'] + '/value')
        # Validate URL
        assert url.scheme == 'https'
        assert url.geturl().startswith(piwebapi_url)

        #  Read the single stream value
        response = requests.get(url.geturl(),
                                auth=security_method, verify=verify_ssl)

        if response.status_code == 200:
            print('{} Snapshot Value'.format(OSI_AF_ATTRIBUTE_TAG))
            print(json.dumps(json.loads(response.text), indent=4, sort_keys=True))
        else:
            print(response.status_code, response.reason, response.text)
    else:
        print(response.status_code, response.reason, response.text)
    return response.status_code


def read_attribute_stream(piwebapi_url, asset_server, user_name, user_password,
                          piwebapi_security_method, verify_ssl):
    """ Read a set of values
        @param piwebapi_url string: The URL of the PI Web API
        @param asset_server string: Name of the Asset Server
        @param user_name string: The user's credentials name
        @param user_password string: The user's credentials password
        @param piwebapi_security_method string: Security method: basic or kerberos
        @param verify_ssl: If certificate verification will be performed
    """
    print('readAttributeStream')

    #  create security method - basic or kerberos
    security_method = call_security_method(
        piwebapi_security_method, user_name, user_password)

    #  Get the sample tag
    request_url = '{}/attributes?path=\\\\{}\\{}\\{}|{}'.format(
        piwebapi_url, asset_server, OSI_AF_DATABASE, OSI_AF_ELEMENT, OSI_AF_ATTRIBUTE_TAG)

    url = urlparse(request_url)
    # Validate URL
    assert url.scheme == 'https'
    assert url.geturl().startswith(piwebapi_url)

    response = requests.get(url.geturl(), auth=security_method, verify=verify_ssl)

    #  Only continue if the first request was successful
    if response.status_code == 200:
        #  Deserialize the JSON Response
        data = json.loads(response.text)

        url = urlparse(piwebapi_url + '/streams/' + data['WebId'] +
                       '/recorded?startTime=*-2d')
        # Validate URL
        assert url.scheme == 'https'
        assert url.geturl().startswith(piwebapi_url)

        #  Read the set of values
        response = requests.get(
            url.geturl(), auth=security_method, verify=verify_ssl)

        if response.status_code == 200:
            print('{} Values'.format(OSI_AF_ATTRIBUTE_TAG))
            print(json.dumps(json.loads(response.text), indent=4, sort_keys=True))
        else:
            print(response.status_code, response.reason, response.text)
    else:
        print(response.status_code, response.reason, response.text)
    return response.status_code


def read_attribute_selected_fields(piwebapi_url, asset_server, user_name, user_password,
                                   piwebapi_security_method, verify_ssl):
    """ Read sampleTag values with selected fields to reduce payload size
        @param piwebapi_url string: The URL of the PI Web API
        @param asset_server string: Name of the Asset Server
        @param user_name string: The user's credentials name
        @param user_password string: The user's credentials password
        @param piwebapi_security_method string: Security method: basic or kerberos
        @param verify_ssl: If certificate verification will be performed
    """
    print('readAttributeSelectedFields')

    #  create security method - basic or kerberos
    security_method = call_security_method(
        piwebapi_security_method, user_name, user_password)

    #  Get the sample tag
    request_url = '{}/attributes?path=\\\\{}\\{}\\{}|{}'.format(
        piwebapi_url, asset_server, OSI_AF_DATABASE, OSI_AF_ELEMENT, OSI_AF_ATTRIBUTE_TAG)
    response = requests.get(request_url,
                            auth=security_method, verify=verify_ssl)

    #  Only continue if the first request was successful
    if response.status_code == 200:
        #  Deserialize the JSON Response
        data = json.loads(response.text)

        url = urlparse(piwebapi_url + '/streams/' + data['WebId'] +
                       '/recorded?startTime=*-2d&selectedFields=Items.Timestamp;Items.Value')
        # Validate URL
        assert url.scheme == 'https'
        assert url.geturl().startswith(piwebapi_url)

        #  Read a set of values and return only the specified columns
        response = requests.get(url.geturl(),
                                auth=security_method, verify=verify_ssl)
        if response.status_code == 200:
            print('SampleTag Values with Selected Fields')
            print(json.dumps(json.loads(response.text), indent=4, sort_keys=True))
        else:
            print(response.status_code, response.reason, response.text)
    else:
        print(response.status_code, response.reason, response.text)
    return response.status_code


def main():
    """ Main method. Receive user input and call the write value methods """
    piwebapi_url = str(input('Enter the PI Web API url: '))
    af_server_name = str(input('Enter the Asset Server Name: '))
    piwebapi_user = str(input('Enter the user name: '))
    piwebapi_password = str(getpass.getpass('Enter the password: '))
    piwebapi_security_method = str(
        input('Enter the security method, Basic or Kerberos: '))
    piwebapi_security_method = piwebapi_security_method.lower()
    verify_ssl_string = str(input('Verify certificates? (Y/N): '))
    
    if (verify_ssl_string.upper() == "N"):
        print('Not verifying certificates poses a security risk and is not recommended')
        verify_ssl = False
    else:
        verify_ssl = True

    read_attribute_snapshot(piwebapi_url, af_server_name, piwebapi_user, piwebapi_password,
                            piwebapi_security_method, verify_ssl)
    read_attribute_stream(piwebapi_url, af_server_name, piwebapi_user, piwebapi_password,
                          piwebapi_security_method, verify_ssl)
    read_attribute_selected_fields(piwebapi_url, af_server_name, piwebapi_user, piwebapi_password,
                                   piwebapi_security_method, verify_ssl)


if __name__ == '__main__':
    main()


batch_call.py
"""This script creates and executes a batch call using the PI Web API
   This python script requires some pre-requisites:
   1.  A back-end server with PI WEB API with CORS enabled.
"""

import json
import getpass
import random
import requests

from requests.auth import HTTPBasicAuth
from requests_kerberos import HTTPKerberosAuth

OSI_AF_DATABASE = 'OSIPythonDatabase'
OSI_AF_ELEMENT = 'OSIPythonElement'
OSI_AF_ATTRIBUTE_TAG = 'OSIPythonAttributeSampleTag'


def call_headers(include_content_type):
    """ Create the header and return a string.
    Parameters:
        include_content_type is a flag that determines whether or not the
        content-type header is included
    """
    if include_content_type is True:
        header = {
            'content-type': 'application/json',
            'X-Requested-With': 'XmlHttpRequest'
        }
    else:
        header = {
            'X-Requested-With': 'XmlHttpRequest'
        }

    return header


def call_security_method(security_method, user_name, user_password):
    """ Create the security method and return a HTTP authorization for basic or kerberos.
    Parameters:
        security_method:  security method to use:  basic or kerberos
        user_name:  User's credentials name
        user_password:  User's credentials password
    """

    if security_method.lower() == 'basic':
        security_auth = HTTPBasicAuth(user_name, user_password)
    else:
        security_auth = HTTPKerberosAuth(mutual_authentication='REQUIRED',
                                         sanitize_mutual_error_response=False)

    return security_auth


def do_batch_call(piwebapiurl, asset_server, user_name, user_password, piwebapi_security_method, piwebapi_verify_certificate):
    """ Create and execute a PI Web API batch call
    Parameters:
        piwebapiurl: the URL of the PI Web API
        assetServer:  the AF server name
        user_name:  User's credentials name
        user_password:  User's credentials password
        piwebapi_security_method:  Security method:  basic or kerberos
        piwebapi_verify_certificate:  If certificate verification will be performed
    """
    print('doBatchCall')

    #  create security method - basic or kerberos
    security_method = call_security_method(
        piwebapi_security_method, user_name, user_password)

    #  Get the sample tag
    request_url = '{}/attributes?path=\\\\{}\\{}\\{}|{}'.format(
        piwebapiurl, asset_server, OSI_AF_DATABASE, OSI_AF_ELEMENT, OSI_AF_ATTRIBUTE_TAG)
    response = requests.get(request_url, auth=security_method, verify=piwebapi_verify_certificate)

    #  Only continue if the first request was successful
    if response.status_code == 200:
        #  Deserialize the JSON Response
        data = json.loads(response.text)

        #  Create the header
        header = call_headers(False)

        #  Create random value to write to a tag.
        attributeValue = '{:.4f}'.format(random.random() * 10)

        #  Create the data for this call
        batch_request = {
            '1': {
                'Method': 'GET',
                'Resource': request_url,
                'Content': '{}'
            },
            '2': {
                'Method': 'GET',
                'Resource': piwebapiurl + '/streams/{0}/value',
                'Content': '{}',
                'Parameters': ['$.1.Content.WebId'],
                'ParentIds': ['1']
            },
            '3': {
                'Method': 'GET',
                'Resource': piwebapiurl + '/streams/{0}/recorded?maxCount=10',
                'Content': '{}',
                'Parameters': ['$.1.Content.WebId'],
                'ParentIds': ['1']
            },
            '4': {
                'Method': 'PUT',
                'Resource': piwebapiurl + '/attributes/{0}/value',
                'Content': '{\'Value\':' + attributeValue + '}',
                'Parameters': ['$.1.Content.WebId'],
                'ParentIds': ['1']
            },
            '5': {
                'Method': 'POST',
                'Resource': piwebapiurl + '/streams/{0}/recorded',
                'Content': '[{\'Value\': \'111\'}, {\'Value\': \'222\'}, {\'Value\': \'333\'}]',
                'Parameters': ['$.1.Content.WebId'],
                'ParentIds': ['1']
            },
            '6': {
                'Method': 'GET',
                'Resource': piwebapiurl + '/streams/{0}/recorded?maxCount=10&selectedFields=Items.Timestamp;Items.Value',
                'Content': '{}',
                'Parameters': ['$.1.Content.WebId'],
                'ParentIds': ['1']
            }
        }

        #  Now that we have the attribute, we need to read the stream value
        response = requests.post(piwebapiurl + '/batch', auth=security_method, verify=piwebapi_verify_certificate,
                                 json=batch_request, headers=header)

        if response.status_code == 207:
            print('Batch Status: ' + str(response.status_code))

            #  Deserialize the JSON Response
            data = json.loads(response.text)

            #  Print the results for each call and format it so it is legible JSON
            print('1: Get the sample tag')
            print(json.dumps(data['1'], indent=4, sort_keys=True))

            print('2: Get the sample tag\'s snapshot value')
            print(json.dumps(data['2'], indent=4, sort_keys=True))

            print('3: Get the sample tag\'s last 10 recorded values')
            print(json.dumps(data['3'], indent=4, sort_keys=True))

            print('4: Write a snapshot value to the sample tag')
            print(json.dumps(data['4'], indent=4, sort_keys=True))

            print('5: Write a set of recorded values to the sample tag')
            print(json.dumps(data['5'], indent=4, sort_keys=True))

            print(
                '6: Get the sample tag\'s last 10 recorded values, only returning the value and timestamp')
            print(json.dumps(data['6'], indent=4, sort_keys=True))

        else:
            print(response.status_code, response.reason, response.text)
    else:
        print(response.status_code, response.reason, response.text)

    return response.status_code


def main():
    """ Main method.  Receive user input and call the do_batch_call method """
    piwebapi_url = str(input('Enter the PI Web API url: '))
    af_server_name = str(input('Enter the Asset Server Name: '))
    piwebapi_user = str(input('Enter the user name: '))
    piwebapi_password = str(getpass.getpass('Enter the password: '))
    piwebapi_security_method = str(
        input('Enter the security method,  Basic or Kerberos:'))
    piwebapi_security_method = piwebapi_security_method.lower()
    piwebapi_verify_certificate = str(input('Verify certificate? (Y/N):'))
    piwebapi_verify_certificate = False if piwebapi_security_method.lower() == 'n' else True


    do_batch_call(piwebapi_url, af_server_name, piwebapi_user, piwebapi_password,
                  piwebapi_security_method, piwebapi_verify_certificate)


if __name__ == '__main__':
    main()
